### PR TITLE
Route53resolver - ResolverRuleAssociation

### PR DESCRIPTION
**Pull request by AmnaIrfan**
### Description of your changes

In this PR we have added the Resolver Rule association resource. This resource falls under the api group "Route53Resolver" which we created in #665 

The previous two resources in #665 (resolverendpoint and resolverrule) were created using codegen. This resource was not available in code gen and was added  manually. 

This resource completes the resolverrule and resolverendpoint resource as it allows users to associate a rule to a vpc.

I have:

- [x]  Read and followed Crossplane's [contribution process].
- [x]  Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested
A sample yaml file can be found under the examples/route53resolver folder. In order to test this resource the user just needs to apply all this file.

Prerequisites: Make sure you set an aws provider, security group, vpc and subnets as well as a resolver endpoint and resolver rule in crossplane before applying this resource.

### CODE REVIEW

**Comment by AmnaIrfan** [comment]
Hey @muvaf -  I was waiting for #665 to be merged. This is now ready for review.
In this PR we add the "resolver rule association" resource to the "route53resolver" api group #665

File: pkg/controller/route53resolver/resolverruleassociation/controller.go [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+		WithOptions(controller.Options{
+			RateLimiter: ratelimiter.NewDefaultManagedRateLimiter(rl),
+		}).
+		For(&v1alpha1.ResolverRuleAssociation{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(v1alpha1.ResolverRuleAssociationGroupVersionKind),
+			managed.WithExternalConnecter(&connector{kube: mgr.GetClient()}),
+			managed.WithReferenceResolver(managed.NewAPISimpleReferenceResolver(mgr.GetClient())),
+			managed.WithConnectionPublishers(),
+			managed.WithInitializers(managed.NewDefaultProviderConfig(mgr.GetClient())),
```
Comment (muvaf) [inline pkg/controller/route53resolver/resolverruleassociation/controller.go:64]:
```suggestion
			managed.WithInitializers(),
```

That initializer is deprecated and our CRDs handle that defaulting now automatically.

File: pkg/controller/route53resolver/resolverruleassociation/controller.go [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+	if meta.GetExternalName(cr) == "" {
+		return managed.ExternalObservation{
+			ResourceExists: false,
+		}, nil
+	}
+	var response *awsr53r.GetResolverRuleAssociationResponse
+	var err error
+
+	response, err = e.client.GetResolverRuleAssociationRequest(&awsr53r.GetResolverRuleAssociationInput{
```
Comment (muvaf) [inline pkg/controller/route53resolver/resolverruleassociation/controller.go:104]:
```suggestion
	response, err := e.client.GetResolverRuleAssociationRequest(&awsr53r.GetResolverRuleAssociationInput{
```

File: pkg/controller/route53resolver/resolverruleassociation/controller.go [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+	current := cr.Spec.ForProvider.DeepCopy()
+
+	cr.SetConditions(xpv1.Available())
+
+	cr.Status.AtProvider = r53r.GenerateRoute53ResolverObservation(*observed)
+
+	return managed.ExternalObservation{
+		ResourceExists:          true,
+		ResourceUpToDate:        true,
+		ResourceLateInitialized: !cmp.Equal(current, &cr.Spec.ForProvider),
```
Comment (muvaf) [inline pkg/controller/route53resolver/resolverruleassociation/controller.go:120]:
```suggestion
```
It seems we don't do any late initialization, so it should be fine to return false here.

File: pkg/controller/route53resolver/resolverruleassociation/controller.go [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+	if err != nil {
+		return managed.ExternalObservation{}, awsclient.Wrap(resource.Ignore(r53r.IsResolverRuleAssociationNotFoundErr, err), errDescribe)
+	}
+
+	observed := response.ResolverRuleAssociation
+	current := cr.Spec.ForProvider.DeepCopy()
+
+	cr.SetConditions(xpv1.Available())
+
+	cr.Status.AtProvider = r53r.GenerateRoute53ResolverObservation(*observed)
```
Comment (muvaf) [inline pkg/controller/route53resolver/resolverruleassociation/controller.go:115]:
```suggestion
	cr.Status.AtProvider = r53r.GenerateRoute53ResolverObservation(*response.ResolverRuleAssociation)
```

File: pkg/controller/route53resolver/resolverruleassociation/controller.go [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+func (e *external) Create(ctx context.Context, mgd resource.Managed) (managed.ExternalCreation, error) {
+	cr, ok := mgd.(*v1alpha1.ResolverRuleAssociation)
+	if !ok {
+		return managed.ExternalCreation{}, errors.New(errUnexpectedObject)
+	}
+
+	cr.Status.SetConditions(xpv1.Creating())
+	if err := e.kube.Status().Update(ctx, cr); err != nil {
+		return managed.ExternalCreation{}, errors.Wrap(err, errStatusUpdate)
+	}
```
Comment (muvaf) [inline pkg/controller/route53resolver/resolverruleassociation/controller.go:133]:
```suggestion
```

We don't need to set conditions for [create](https://github.com/crossplane/crossplane-runtime/blob/master/pkg/reconciler/managed/reconciler.go#L714) and [delete](https://github.com/crossplane/crossplane-runtime/blob/master/pkg/reconciler/managed/reconciler.go#L539) anymore.

File: pkg/controller/route53resolver/resolverruleassociation/controller.go [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+	return managed.ExternalUpdate{}, nil
+}
+
+func (e *external) Delete(ctx context.Context, mgd resource.Managed) error {
+	cr, ok := mgd.(*v1alpha1.ResolverRuleAssociation)
+	if !ok {
+		return errors.New(errUnexpectedObject)
+	}
+
+	cr.Status.SetConditions(xpv1.Deleting())
```
Comment (muvaf) [inline pkg/controller/route53resolver/resolverruleassociation/controller.go:158]:
```suggestion
```

File: pkg/controller/route53resolver/resolverruleassociation/controller.go [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+			ResourceExists: false,
+		}, nil
+	}
+	var response *awsr53r.GetResolverRuleAssociationResponse
+	var err error
+
+	response, err = e.client.GetResolverRuleAssociationRequest(&awsr53r.GetResolverRuleAssociationInput{
+		ResolverRuleAssociationId: aws.String(meta.GetExternalName(cr))}).Send(ctx)
+	if err != nil {
+		return managed.ExternalObservation{}, awsclient.Wrap(resource.Ignore(r53r.IsResolverRuleAssociationNotFoundErr, err), errDescribe)
```
Comment (muvaf) [inline pkg/controller/route53resolver/resolverruleassociation/controller.go:107]:
```suggestion
		return managed.ExternalObservation{}, awsclient.Wrap(err, errDescribe)
```

If the ID is given via external name annotation and the association is not found, we'd normally expect to create an association with that ID. But from what I can tell, association ID is assigned by AWS and cannot be affected by us. So, ignoring `NotFound` will cause us to create a new resource and override the given ID, which is probably not something user wants if they give the ID. So, ignoring the error here makes more sense if the identifier we use can be decided by us. I know some resources similar to this one do ignore but that's due to historical reasons.

**Review by muvaf — COMMENTED** [review]
I have only a few small comments to be adressed. Thank you @AmnaIrfan !

**Comment by AmnaIrfan** [comment]
@muvaf  Thank you for the review.  Please let me know if I need to do anymore changes. Thank you!

**Review by AaronME — CHANGES_REQUESTED** [review]
@AmnaIrfan Thank you for the PR!

I was unable to get a working build from the branch until I removed `errStatusUpdate.` With that removed it builds, but I didn't get a successful creation of a resolverruleassociation resource.

To get this approved we need:
- Rebase on current `master`
- Add `route53resolver` to the list of GENERATED_SERVICES in the [Makefile](https://github.com/crossplane/provider-aws/blob/988e66c6befcb6b6e7ce575d56f8254683aa442b/Makefile#L10)
- Run `make reviewable` before pushing.

Thanks, again!

**Comment by haarchri** [comment]
i was not aware of this PR i added the same resource https://github.com/crossplane/provider-aws/pull/968 - i can close my PR if we go to merge this one

**Comment by haarchri** [comment]
implemented via https://github.com/crossplane/provider-aws/pull/968

**Pull request closed without merging**